// @nimblebrain/synapse/dist/chunk-HLT5UBJF.cjs
'use strict';

// src/codegen/type-generator.ts
function generateTypes(tools, appName) {
  const lines = [];
  const mapName = `${toPascalCase(appName)}ToolMap`;
  lines.push("/**");
  lines.push(` * Auto-generated by @nimblebrain/synapse codegen`);
  lines.push(` * Source: ${appName}`);
  lines.push(` *`);
  lines.push(
    ` * DO NOT EDIT \u2014 regenerate with: npx synapse codegen --from-manifest ./manifest.json`
  );
  lines.push(` */`);
  lines.push("");
  const mapEntries = [];
  for (const tool of tools) {
    const baseName = toPascalCase(tool.name);
    const inputName = `${baseName}Input`;
    lines.push(schemaToInterface(inputName, tool.inputSchema));
    lines.push("");
    let outputName = "unknown";
    if (tool.outputSchema) {
      outputName = `${baseName}Output`;
      lines.push(schemaToInterface(outputName, tool.outputSchema));
      lines.push("");
    }
    mapEntries.push(`  ${tool.name}: { input: ${inputName}; output: ${outputName} };`);
  }
  lines.push(`/** Tool type map for typed useCallTool */`);
  lines.push(`export interface ${mapName} {`);
  for (const entry of mapEntries) {
    lines.push(entry);
  }
  lines.push(`}`);
  lines.push("");
  return lines.join("\n");
}
function schemaToInterface(name, schema) {
  const type = schemaToType(schema, name);
  if (schema.type === "object" || schema.properties) {
    return generateInterface(name, schema);
  }
  return `export type ${name} = ${type};`;
}
function generateInterface(name, schema) {
  const lines = [];
  lines.push(`export interface ${name} {`);
  const props = schema.properties ?? {};
  const required = new Set(schema.required ?? []);
  for (const [key, propSchema] of Object.entries(props)) {
    const isRequired = required.has(key);
    const type = schemaToType(propSchema, name + toPascalCase(key));
    const desc = propSchema.description;
    if (desc) {
      lines.push(`  /** ${desc} */`);
    }
    lines.push(`  ${key}${isRequired ? "" : "?"}: ${type};`);
  }
  lines.push(`}`);
  return lines.join("\n");
}
function schemaToType(schema, context) {
  if (schema.enum && Array.isArray(schema.enum)) {
    return schema.enum.map((v) => typeof v === "string" ? `"${v}"` : String(v)).join(" | ");
  }
  if (schema.oneOf && Array.isArray(schema.oneOf)) {
    return schema.oneOf.map((s, i) => schemaToType(s, `${context}Option${i}`)).join(" | ");
  }
  if (schema.anyOf && Array.isArray(schema.anyOf)) {
    return schema.anyOf.map((s, i) => schemaToType(s, `${context}Option${i}`)).join(" | ");
  }
  if (schema.allOf && Array.isArray(schema.allOf)) {
    return schema.allOf.map((s, i) => schemaToType(s, `${context}Part${i}`)).join(" & ");
  }
  const type = schema.type;
  if (Array.isArray(type)) {
    return type.map((t) => primitiveType(t)).join(" | ");
  }
  switch (type) {
    case "string":
      return "string";
    case "number":
    case "integer":
      return "number";
    case "boolean":
      return "boolean";
    case "null":
      return "null";
    case "array": {
      const items = schema.items;
      if (items) {
        return `${schemaToType(items, `${context}Item`)}[]`;
      }
      return "unknown[]";
    }
    case "object": {
      if (schema.properties) {
        const props = schema.properties;
        const required = new Set(schema.required ?? []);
        const fields = Object.entries(props).map(([key, propSchema]) => {
          const t = schemaToType(propSchema, context + toPascalCase(key));
          return `${key}${required.has(key) ? "" : "?"}: ${t}`;
        }).join("; ");
        return `{ ${fields} }`;
      }
      return "Record<string, unknown>";
    }
    default:
      return "unknown";
  }
}
function primitiveType(t) {
  switch (t) {
    case "string":
      return "string";
    case "number":
    case "integer":
      return "number";
    case "boolean":
      return "boolean";
    case "null":
      return "null";
    default:
      return "unknown";
  }
}
function toPascalCase(str) {
  return str.split(/[-_@/]/).filter(Boolean).map((word) => word.charAt(0).toUpperCase() + word.slice(1)).join("");
}

exports.generateTypes = generateTypes;
