// @nimblebrain/synapse/dist/index.cjs
'use strict';

var chunk54YIV4ZL_cjs = require('./chunk-54YIV4ZL.cjs');

// src/store.ts
function createStore(synapse, config) {
  let state = structuredClone(config.initialState);
  const subscribers = /* @__PURE__ */ new Set();
  let destroyed = false;
  let persistTimer = null;
  const dispatch = {};
  for (const key of Object.keys(config.actions)) {
    dispatch[key] = (payload) => {
      if (destroyed) return;
      state = config.actions[key](state, payload);
      notify();
    };
  }
  function notify() {
    for (const cb of subscribers) cb(state);
    if (config.visibleToAgent) pushToAgent();
    if (config.persist) schedulePersist();
  }
  function pushToAgent() {
    const summary = config.summarize?.(state);
    synapse.setVisibleState(state, summary);
  }
  function schedulePersist() {
    if (persistTimer) clearTimeout(persistTimer);
    persistTimer = setTimeout(() => {
      synapse._request("synapse/persist-state", {
        state,
        version: config.version
      }).catch(() => {
      });
      persistTimer = null;
    }, 500);
  }
  let unsubStateLoaded;
  if (config.persist) {
    unsubStateLoaded = synapse._onMessage("synapse/state-loaded", (params) => {
      if (!params?.state) return;
      let loaded = params.state;
      const loadedVersion = params.version ?? 1;
      const currentVersion = config.version ?? 1;
      if (config.migrations && loadedVersion < currentVersion) {
        const startIdx = loadedVersion - 1;
        for (let i = startIdx; i < config.migrations.length; i++) {
          loaded = config.migrations[i](loaded);
        }
      }
      store.hydrate(loaded);
    });
  }
  const store = {
    getState() {
      return state;
    },
    subscribe(callback) {
      subscribers.add(callback);
      return () => {
        subscribers.delete(callback);
      };
    },
    dispatch,
    hydrate(newState) {
      state = newState;
      for (const cb of subscribers) cb(state);
    },
    destroy() {
      if (destroyed) return;
      destroyed = true;
      if (persistTimer) clearTimeout(persistTimer);
      subscribers.clear();
      unsubStateLoaded?.();
    }
  };
  return store;
}

Object.defineProperty(exports, "connect", {
  enumerable: true,
  get: function () { return chunk54YIV4ZL_cjs.connect; }
});
Object.defineProperty(exports, "createSynapse", {
  enumerable: true,
  get: function () { return chunk54YIV4ZL_cjs.createSynapse; }
});
exports.createStore = createStore;
