// @nimblebrain/synapse/dist/vite/index.cjs
'use strict';

var child_process = require('child_process');
var fs = require('fs');
var path = require('path');

// src/vite/plugin.ts
function synapseVite(options = {}) {
  const enablePreview = options.preview !== false;
  let manifest = null;
  let appName = options.appName ?? "app";
  let serverProcess = null;
  const pendingRequests = /* @__PURE__ */ new Map();
  let serverBuffer = "";
  function loadManifest(root) {
    const manifestPath = options.manifest ? path.resolve(options.manifest) : path.resolve(root, "..", "manifest.json");
    if (!fs.existsSync(manifestPath)) return null;
    try {
      return JSON.parse(fs.readFileSync(manifestPath, "utf-8"));
    } catch {
      return null;
    }
  }
  function deriveServerCmd(m, root) {
    if (options.serverCmd) return options.serverCmd;
    const cfg = m.server?.mcp_config;
    if (!cfg?.command) return null;
    const serverDir = path.resolve(root, "..");
    let cmd = cfg.command;
    const args = cfg.args ?? [];
    if (cmd === "python" && fs.existsSync(path.join(serverDir, "pyproject.toml"))) {
      cmd = "uv run python";
    }
    return `cd ${JSON.stringify(serverDir)} && ${cmd} ${args.join(" ")}`;
  }
  function startServer(cmd) {
    serverProcess = child_process.spawn(cmd, {
      shell: true,
      stdio: ["pipe", "pipe", "pipe"]
    });
    serverProcess.stderr?.on("data", (d) => {
      process.stderr.write(`  [mcp] ${d}`);
    });
    serverProcess.stdout?.on("data", (d) => {
      serverBuffer += d.toString();
      const lines = serverBuffer.split("\n");
      serverBuffer = lines.pop() ?? "";
      for (const line of lines) {
        if (!line.trim()) continue;
        try {
          const msg = JSON.parse(line);
          if (msg.id && pendingRequests.has(msg.id)) {
            const p = pendingRequests.get(msg.id);
            pendingRequests.delete(msg.id);
            p?.resolve(msg);
          }
        } catch {
          process.stderr.write(`  [mcp] ${line}
`);
        }
      }
    });
    serverProcess.on("exit", (code) => {
      if (code !== null && code !== 0) {
        console.error(`  [mcp] Server exited with code ${code}`);
      }
      serverProcess = null;
    });
    sendToServer({
      jsonrpc: "2.0",
      id: "init-1",
      method: "initialize",
      params: {
        protocolVersion: "2024-11-05",
        capabilities: {},
        clientInfo: { name: "synapse-preview", version: "0.1.0" }
      }
    });
  }
  function sendToServer(msg) {
    if (!serverProcess?.stdin?.writable) return;
    serverProcess.stdin.write(`${JSON.stringify(msg)}
`);
  }
  function callServerTool(name, args) {
    return new Promise((resolve2, reject) => {
      const id = `preview-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
      pendingRequests.set(id, { resolve: resolve2, reject });
      sendToServer({
        jsonrpc: "2.0",
        id,
        method: "tools/call",
        params: { name, arguments: args }
      });
      setTimeout(() => {
        if (pendingRequests.has(id)) {
          pendingRequests.delete(id);
          reject(new Error("Tool call timed out (10s)"));
        }
      }, 1e4);
    });
  }
  return {
    name: "synapse",
    config() {
      return {
        define: {
          "import.meta.env.SYNAPSE_APP_NAME": JSON.stringify(appName)
        },
        server: {
          hmr: {
            protocol: "ws",
            host: "localhost"
          }
        }
      };
    },
    configResolved(config) {
      manifest = loadManifest(config.root);
      if (manifest?.name) {
        appName = options.appName ?? manifest.name;
      }
    },
    configureServer(server) {
      if (enablePreview && manifest) {
        const cmd = deriveServerCmd(manifest, server.config.root);
        if (cmd) {
          console.log(`
  [synapse] Starting MCP server: ${cmd}
`);
          startServer(cmd);
        }
      }
      server.middlewares.use((req, res, next) => {
        res.setHeader("Access-Control-Allow-Origin", "*");
        res.setHeader("Access-Control-Allow-Methods", "*");
        res.setHeader("Access-Control-Allow-Headers", "*");
        if (req.url === "/__preview" || req.url === "/__preview/") {
          res.writeHead(200, { "Content-Type": "text/html" });
          res.end(previewHostHtml(appName));
          return;
        }
        if (req.method === "POST" && req.url === "/__mcp") {
          let body = "";
          req.on("data", (chunk) => {
            body += chunk.toString();
          });
          req.on("end", async () => {
            try {
              const msg = JSON.parse(body);
              const result = await callServerTool(msg.params.name, msg.params.arguments || {});
              res.writeHead(200, { "Content-Type": "application/json" });
              res.end(JSON.stringify(result));
            } catch (err) {
              res.writeHead(200, { "Content-Type": "application/json" });
              res.end(
                JSON.stringify({
                  jsonrpc: "2.0",
                  id: JSON.parse(body).id,
                  error: { code: -32e3, message: err.message }
                })
              );
            }
          });
          return;
        }
        next();
      });
    },
    buildEnd() {
      if (serverProcess) {
        serverProcess.kill("SIGTERM");
        serverProcess = null;
      }
    }
  };
}
function previewHostHtml(appName) {
  return `<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8" />
  <title>${appName} \u2014 Synapse Preview</title>
  <style>
    * { margin: 0; padding: 0; box-sizing: border-box; }
    body { font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif; background: #0f172a; color: #e2e8f0; }
    header { padding: 10px 16px; background: #1e293b; border-bottom: 1px solid #334155; display: flex; align-items: center; gap: 10px; font-size: 13px; }
    header .dot { width: 8px; height: 8px; border-radius: 50%; background: #22c55e; }
    header .name { font-weight: 600; }
    header .spacer { flex: 1; }
    header button { background: #334155; border: none; color: #e2e8f0; padding: 3px 8px; border-radius: 4px; cursor: pointer; font-size: 12px; }
    header .url { color: #64748b; font-size: 11px; font-family: monospace; }
    iframe { width: 100%; height: calc(100vh - 41px); border: none; }
  </style>
</head>
<body>
  <header>
    <span class="dot"></span>
    <span class="name">${appName}</span>
    <span class="spacer"></span>
    <button id="toggle">Toggle Theme</button>
    <span class="url">Synapse Preview</span>
  </header>
  <iframe id="app"></iframe>

  <script>
    var iframe = document.getElementById("app");
    var dark = true;

    function getTokens(d) {
      return d ? {
        "--color-background-primary":"#0f172a","--color-text-primary":"#e2e8f0",
        "--color-background-secondary":"#1e293b","--color-text-primary":"#e2e8f0",
        "--color-text-accent":"#6366f1","--nb-color-accent-foreground":"#fff",
        "--color-text-secondary":"#94a3b8","--color-border-primary":"#334155",
        "--color-ring-primary":"#6366f1","--nb-color-danger":"#ef4444",
        "--border-radius-sm":"0.5rem","--font-sans":"-apple-system,BlinkMacSystemFont,sans-serif"
      } : {
        "--color-background-primary":"#ffffff","--color-text-primary":"#0f172a",
        "--color-background-secondary":"#f8fafc","--color-text-primary":"#0f172a",
        "--color-text-accent":"#6366f1","--nb-color-accent-foreground":"#fff",
        "--color-text-secondary":"#64748b","--color-border-primary":"#e2e8f0",
        "--color-ring-primary":"#6366f1","--nb-color-danger":"#ef4444",
        "--border-radius-sm":"0.5rem","--font-sans":"-apple-system,BlinkMacSystemFont,sans-serif"
      };
    }

    function post(msg) { iframe.contentWindow.postMessage(msg, "*"); }

    window.addEventListener("message", async function(e) {
      if (e.source !== iframe.contentWindow) return;
      var msg = e.data;
      if (!msg || typeof msg !== "object") return;

      // ext-apps handshake
      if (msg.method === "ui/initialize" && msg.id) {
        post({ jsonrpc:"2.0", id:msg.id, result: {
          protocolVersion:"2026-01-26",
          serverInfo:{name:"nimblebrain",version:"preview"},
          capabilities:{openLinks:{},serverTools:{}},
          hostContext:{theme:dark?"dark":"light",primaryColor:"#6366f1",tokens:getTokens(dark)}
        }});
        return;
      }
      if (msg.method === "ui/notifications/initialized") return;

      // Tool calls \u2014 proxy via Vite middleware
      if (msg.method === "tools/call" && msg.id) {
        var originalId = msg.id;
        try {
          var r = await fetch("/__mcp", {
            method:"POST", headers:{"Content-Type":"application/json"},
            body: JSON.stringify({jsonrpc:"2.0",id:msg.id,method:"tools/call",params:{name:msg.params.name,arguments:msg.params.arguments||{}}})
          });
          var response = await r.json();
          response.id = originalId;
          post(response);
          // Notify the app that data changed so hooks auto-refresh
          // Only for mutating operations (not list/search/get which are read-only)
          var tn = msg.params.name || "";
          if (!response.error && !tn.startsWith("list_") && !tn.startsWith("search_") && !tn.startsWith("get_") && !tn.startsWith("query_")) {
            post({jsonrpc:"2.0",method:"synapse/data-changed",params:{tool:tn,server:"preview"}});
          }
        } catch(err) {
          post({jsonrpc:"2.0",id:originalId,error:{code:-32000,message:err.message}});
        }
        return;
      }

      // Log other messages
      if (msg.method === "synapse/chat") console.log("[chat]", msg.params?.message);
      else if (msg.method === "synapse/action") console.log("[action]", msg.params?.action, msg.params);
      else if (msg.method === "ui/update-model-context") { console.log("[model-context]", msg.params?.structuredContent); if (msg.id) post({jsonrpc:"2.0",id:msg.id,result:{}}); }
      else if (msg.method === "synapse/keydown") { /* ignore */ }
      else if (msg.method) console.log("[bridge]", msg.method, msg);
    });

    document.getElementById("toggle").onclick = function() {
      dark = !dark;
      document.body.style.background = dark ? "#0f172a" : "#f1f5f9";
      post({jsonrpc:"2.0",method:"synapse/theme-changed",params:{mode:dark?"dark":"light",tokens:getTokens(dark)}});
    };

    // Load the iframe AFTER the message listener is attached to avoid
    // a race where the app sends ui/initialize before the bridge is ready.
    iframe.src = "/";
  </script>
</body>
</html>`;
}

exports.synapseVite = synapseVite;
