// @nimblebrain/synapse/dist/react/index.cjs
'use strict';

var chunk54YIV4ZL_cjs = require('../chunk-54YIV4ZL.cjs');
var react = require('react');
var jsxRuntime = require('react/jsx-runtime');

var AppContext = react.createContext(null);
function AppProvider({ children, name, version, autoResize }) {
  const [app, setApp] = react.useState(null);
  const connectingRef = react.useRef(false);
  react.useEffect(() => {
    if (connectingRef.current) return;
    connectingRef.current = true;
    chunk54YIV4ZL_cjs.connect({ name, version, autoResize }).then((a) => {
      setApp(a);
    });
  }, []);
  if (!app) return null;
  return /* @__PURE__ */ jsxRuntime.jsx(AppContext.Provider, { value: app, children });
}
function useAppContext() {
  const ctx = react.useContext(AppContext);
  if (!ctx) {
    throw new Error(
      "useApp must be used within an <AppProvider>. Wrap your component tree with <AppProvider>."
    );
  }
  return ctx;
}

// src/react/connect-hooks.ts
function useApp() {
  return useAppContext();
}
function useToolResult() {
  const app = useAppContext();
  const [data, setData] = react.useState(null);
  react.useEffect(() => {
    return app.on("tool-result", (result) => {
      setData(result);
    });
  }, [app]);
  return data;
}
function useToolInput() {
  const app = useAppContext();
  const [input, setInput] = react.useState(null);
  react.useEffect(() => {
    return app.on("tool-input", (args) => {
      setInput(args);
    });
  }, [app]);
  return input;
}
function useConnectTheme() {
  const app = useAppContext();
  const [theme, setTheme] = react.useState(() => app.theme);
  react.useEffect(() => {
    setTheme(app.theme);
    return app.on("theme-changed", (t) => {
      setTheme(t);
    });
  }, [app]);
  return theme;
}
function useResize() {
  const app = useAppContext();
  return react.useCallback((width, height) => app.resize(width, height), [app]);
}
var SynapseContext = react.createContext(null);
function SynapseProvider({ children, ...options }) {
  const ref = react.useRef(null);
  if (ref.current === null || ref.current.destroyed) {
    ref.current = chunk54YIV4ZL_cjs.createSynapse(options);
  }
  return /* @__PURE__ */ jsxRuntime.jsxs(SynapseContext.Provider, { value: ref.current, children: [
    /* @__PURE__ */ jsxRuntime.jsx(ThemeInjector, { synapse: ref.current }),
    children
  ] });
}
function ThemeInjector({ synapse }) {
  const [theme, setTheme] = react.useState(() => synapse.getTheme());
  react.useEffect(() => {
    setTheme(synapse.getTheme());
    return synapse.onThemeChanged(setTheme);
  }, [synapse]);
  react.useEffect(() => {
    if (theme.tokens) {
      const style = document.documentElement.style;
      for (const [k, v] of Object.entries(theme.tokens)) {
        style.setProperty(k, v);
      }
    }
  }, [theme]);
  return null;
}
function useSynapseContext() {
  const ctx = react.useContext(SynapseContext);
  if (!ctx) {
    throw new Error(
      "useSynapse must be used within a <SynapseProvider>. Wrap your app component tree with <SynapseProvider>."
    );
  }
  return ctx;
}

// src/react/hooks.ts
function useSynapse() {
  return useSynapseContext();
}
function useCallTool(toolName) {
  const synapse = useSynapseContext();
  const [isPending, setIsPending] = react.useState(false);
  const [error, setError] = react.useState(null);
  const [data, setData] = react.useState(null);
  const callIdRef = react.useRef(0);
  const call = react.useCallback(
    async (args) => {
      const id = ++callIdRef.current;
      setIsPending(true);
      setError(null);
      try {
        const result = await synapse.callTool(toolName, args);
        if (id === callIdRef.current) {
          setData(result.data);
          setIsPending(false);
        }
        return result;
      } catch (err) {
        if (id === callIdRef.current) {
          const e = err instanceof Error ? err : new Error(String(err));
          setError(e);
          setIsPending(false);
        }
        throw err;
      }
    },
    [synapse, toolName]
  );
  return { call, isPending, error, data };
}
function useDataSync(callback) {
  const synapse = useSynapseContext();
  const callbackRef = react.useRef(callback);
  callbackRef.current = callback;
  react.useEffect(() => {
    return synapse.onDataChanged((event) => callbackRef.current(event));
  }, [synapse]);
}
function useAgentAction(callback) {
  const synapse = useSynapseContext();
  const callbackRef = react.useRef(callback);
  callbackRef.current = callback;
  react.useEffect(() => {
    return synapse.onAction((action) => callbackRef.current(action));
  }, [synapse]);
}
function useTheme() {
  const synapse = useSynapseContext();
  const [theme, setTheme] = react.useState(() => synapse.getTheme());
  react.useEffect(() => {
    setTheme(synapse.getTheme());
    return synapse.onThemeChanged(setTheme);
  }, [synapse]);
  return theme;
}
function useAction() {
  const synapse = useSynapseContext();
  return react.useCallback(
    (action, params) => synapse.action(action, params),
    [synapse]
  );
}
function useChat() {
  const synapse = useSynapseContext();
  return react.useCallback(
    (message, context) => synapse.chat(message, context),
    [synapse]
  );
}
function useVisibleState(factory, deps) {
  const synapse = useSynapseContext();
  const push = react.useCallback(
    (state, summary) => synapse.setVisibleState(state, summary),
    [synapse]
  );
  const factoryRef = react.useRef(factory);
  factoryRef.current = factory;
  react.useEffect(() => {
    if (!factoryRef.current) return;
    const { state, summary } = factoryRef.current();
    push(state, summary);
  }, [...deps ?? [], push]);
  if (!factory) return push;
}
function useFileUpload() {
  const synapse = useSynapseContext();
  const [isPending, setIsPending] = react.useState(false);
  const pickFile = react.useCallback(
    async (options) => {
      setIsPending(true);
      try {
        return await synapse.pickFile(options);
      } finally {
        setIsPending(false);
      }
    },
    [synapse]
  );
  const pickFiles = react.useCallback(
    async (options) => {
      setIsPending(true);
      try {
        return await synapse.pickFiles(options);
      } finally {
        setIsPending(false);
      }
    },
    [synapse]
  );
  return { pickFile, pickFiles, isPending };
}
function useStore(store) {
  const state = react.useSyncExternalStore(
    (onStoreChange) => store.subscribe(onStoreChange),
    () => store.getState(),
    () => store.getState()
  );
  return { state, dispatch: store.dispatch };
}

exports.AppProvider = AppProvider;
exports.SynapseProvider = SynapseProvider;
exports.useAction = useAction;
exports.useAgentAction = useAgentAction;
exports.useApp = useApp;
exports.useCallTool = useCallTool;
exports.useChat = useChat;
exports.useConnectTheme = useConnectTheme;
exports.useDataSync = useDataSync;
exports.useFileUpload = useFileUpload;
exports.useResize = useResize;
exports.useStore = useStore;
exports.useSynapse = useSynapse;
exports.useTheme = useTheme;
exports.useToolInput = useToolInput;
exports.useToolResult = useToolResult;
exports.useVisibleState = useVisibleState;
