// @nimblebrain/synapse/dist/react/index.js
import { connect, createSynapse } from '../chunk-LYJHA5B2.js';
import { createContext, useState, useRef, useEffect, useCallback, useSyncExternalStore, useContext } from 'react';
import { jsx, jsxs } from 'react/jsx-runtime';

var AppContext = createContext(null);
function AppProvider({ children, name, version, autoResize }) {
  const [app, setApp] = useState(null);
  const connectingRef = useRef(false);
  useEffect(() => {
    if (connectingRef.current) return;
    connectingRef.current = true;
    connect({ name, version, autoResize }).then((a) => {
      setApp(a);
    });
  }, []);
  if (!app) return null;
  return /* @__PURE__ */ jsx(AppContext.Provider, { value: app, children });
}
function useAppContext() {
  const ctx = useContext(AppContext);
  if (!ctx) {
    throw new Error(
      "useApp must be used within an <AppProvider>. Wrap your component tree with <AppProvider>."
    );
  }
  return ctx;
}

// src/react/connect-hooks.ts
function useApp() {
  return useAppContext();
}
function useToolResult() {
  const app = useAppContext();
  const [data, setData] = useState(null);
  useEffect(() => {
    return app.on("tool-result", (result) => {
      setData(result);
    });
  }, [app]);
  return data;
}
function useToolInput() {
  const app = useAppContext();
  const [input, setInput] = useState(null);
  useEffect(() => {
    return app.on("tool-input", (args) => {
      setInput(args);
    });
  }, [app]);
  return input;
}
function useConnectTheme() {
  const app = useAppContext();
  const [theme, setTheme] = useState(() => app.theme);
  useEffect(() => {
    setTheme(app.theme);
    return app.on("theme-changed", (t) => {
      setTheme(t);
    });
  }, [app]);
  return theme;
}
function useResize() {
  const app = useAppContext();
  return useCallback((width, height) => app.resize(width, height), [app]);
}
var SynapseContext = createContext(null);
function SynapseProvider({ children, ...options }) {
  const ref = useRef(null);
  if (ref.current === null || ref.current.destroyed) {
    ref.current = createSynapse(options);
  }
  return /* @__PURE__ */ jsxs(SynapseContext.Provider, { value: ref.current, children: [
    /* @__PURE__ */ jsx(ThemeInjector, { synapse: ref.current }),
    children
  ] });
}
function ThemeInjector({ synapse }) {
  const [theme, setTheme] = useState(() => synapse.getTheme());
  useEffect(() => {
    setTheme(synapse.getTheme());
    return synapse.onThemeChanged(setTheme);
  }, [synapse]);
  useEffect(() => {
    if (theme.tokens) {
      const style = document.documentElement.style;
      for (const [k, v] of Object.entries(theme.tokens)) {
        style.setProperty(k, v);
      }
    }
  }, [theme]);
  return null;
}
function useSynapseContext() {
  const ctx = useContext(SynapseContext);
  if (!ctx) {
    throw new Error(
      "useSynapse must be used within a <SynapseProvider>. Wrap your app component tree with <SynapseProvider>."
    );
  }
  return ctx;
}

// src/react/hooks.ts
function useSynapse() {
  return useSynapseContext();
}
function useCallTool(toolName) {
  const synapse = useSynapseContext();
  const [isPending, setIsPending] = useState(false);
  const [error, setError] = useState(null);
  const [data, setData] = useState(null);
  const callIdRef = useRef(0);
  const call = useCallback(
    async (args) => {
      const id = ++callIdRef.current;
      setIsPending(true);
      setError(null);
      try {
        const result = await synapse.callTool(toolName, args);
        if (id === callIdRef.current) {
          setData(result.data);
          setIsPending(false);
        }
        return result;
      } catch (err) {
        if (id === callIdRef.current) {
          const e = err instanceof Error ? err : new Error(String(err));
          setError(e);
          setIsPending(false);
        }
        throw err;
      }
    },
    [synapse, toolName]
  );
  return { call, isPending, error, data };
}
function useDataSync(callback) {
  const synapse = useSynapseContext();
  const callbackRef = useRef(callback);
  callbackRef.current = callback;
  useEffect(() => {
    return synapse.onDataChanged((event) => callbackRef.current(event));
  }, [synapse]);
}
function useAgentAction(callback) {
  const synapse = useSynapseContext();
  const callbackRef = useRef(callback);
  callbackRef.current = callback;
  useEffect(() => {
    return synapse.onAction((action) => callbackRef.current(action));
  }, [synapse]);
}
function useTheme() {
  const synapse = useSynapseContext();
  const [theme, setTheme] = useState(() => synapse.getTheme());
  useEffect(() => {
    setTheme(synapse.getTheme());
    return synapse.onThemeChanged(setTheme);
  }, [synapse]);
  return theme;
}
function useAction() {
  const synapse = useSynapseContext();
  return useCallback(
    (action, params) => synapse.action(action, params),
    [synapse]
  );
}
function useChat() {
  const synapse = useSynapseContext();
  return useCallback(
    (message, context) => synapse.chat(message, context),
    [synapse]
  );
}
function useVisibleState(factory, deps) {
  const synapse = useSynapseContext();
  const push = useCallback(
    (state, summary) => synapse.setVisibleState(state, summary),
    [synapse]
  );
  const factoryRef = useRef(factory);
  factoryRef.current = factory;
  useEffect(() => {
    if (!factoryRef.current) return;
    const { state, summary } = factoryRef.current();
    push(state, summary);
  }, [...deps ?? [], push]);
  if (!factory) return push;
}
function useFileUpload() {
  const synapse = useSynapseContext();
  const [isPending, setIsPending] = useState(false);
  const pickFile = useCallback(
    async (options) => {
      setIsPending(true);
      try {
        return await synapse.pickFile(options);
      } finally {
        setIsPending(false);
      }
    },
    [synapse]
  );
  const pickFiles = useCallback(
    async (options) => {
      setIsPending(true);
      try {
        return await synapse.pickFiles(options);
      } finally {
        setIsPending(false);
      }
    },
    [synapse]
  );
  return { pickFile, pickFiles, isPending };
}
function useStore(store) {
  const state = useSyncExternalStore(
    (onStoreChange) => store.subscribe(onStoreChange),
    () => store.getState(),
    () => store.getState()
  );
  return { state, dispatch: store.dispatch };
}

export { AppProvider, SynapseProvider, useAction, useAgentAction, useApp, useCallTool, useChat, useConnectTheme, useDataSync, useFileUpload, useResize, useStore, useSynapse, useTheme, useToolInput, useToolResult, useVisibleState };
